### ISSUE TITLE
Using the wdio runner how do I force a new session between tests instead of between files?

### ISSUE THREAD

**Issue by andyboyne** [webdriverio/webdriverio#1248]
I'm using WebDriverIO to test against OpenFin (a chrome container - https://openfin.co/). Between tests I need to explicitly shut the container down.  Having more than one test in a file cause problems because WebDriverIO is not initializing the client between the tests, only between files.  How can I control when the client is created?

**Comment by christian-bromann** [comment]
There is no way yet to recreate a session besides splitting tests up in multiple test files.

**Comment by andyboyne** [comment]
Ok, thanks for the quick response. Is there anything I could hack quickly to overcome this, or is it a deep change?

**Comment by christian-bromann** [comment]
it's a deep change, I was thinking of having a reload function that reloads a session but I am not sure. Why is it a problem to split up tests into multiple files?

**Comment by andyboyne** [comment]
There will be hundreds/thousands of tests.  Usually I would have something along the lines of:

// feature-a.js
describe('..........', function() {
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
}

// feature-b.js
describe('..........', function() {
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
    it('should.........', function(.....
}

Having to put each 'it' into its own file will a bit much no?

**Comment by christian-bromann** [comment]
Agree, but why do you need a complete new session everytime?

**Comment by andyboyne** [comment]
https://github.com/openfin/hello-openfin-selenium-example

"Since all HTML5 applications in the OpenFin environment need to be started with OpenFin API, chromeDriver.get(URL) is not supported. Test code needs to provide the OpenFinRVM and a HTML5 app configuration to Chromedriver so it can start OpenFin Runtime and the html5 application before tests can be run. config.js has the location of OpenFinRVM and the URL for the 'Hello OpenFin' configuration file."

and 

"Since the OpenFin Runtime is started by OpenFinRVM, Chromedriver does not have direct control of the OpenFin Runtime. Chromedriver must be started before any test runs. Once a test is complete, it needs to shut down OpenFin Runtime by running javascript code "fin.desktop.System.exit();". driver.quit() does not shut down OpenFin Runtime since it does not have access."


Basically we provide config like:

        chromeOptions: {
            debuggerAddress: 'localhost:9120',
            binary: "start-openfin.bat",       <---- this actually starts the application
       }

.. and the chromedriver then attaches to OpenFin.  Since we have to manually teardown the container, the session never re-establishes meaning only one test per file :o(

**Comment by andyboyne** [comment]
How much work is it to do the reloadSession - can you describe roughly what this would entail?

**Comment by aboutlo** [comment]
I guess I got the same issue with cucumber & webdriver

```
Background:
    Given I visit "/foo" 
    Then I log as "user@domain.tld" with password "pswd"
    Then I click ".action.login"

  Scenario: Do something
    Then I visit '/foo/page1'
    Then I should count "3" elements of ".foo"

  Scenario: Do something else
    Then I visit '/foo/page2'
    Then I should count "2" elements of ".bar"
```

I expect between each scenario the session is reloaded. Am I wrong?

**Comment by christian-bromann** [comment]
@aboutlo yeah, in WebdriverIO you run one session per feature file (that contains multiple scenarios).

@andyboyne never heart of OpenFin. It sounds like this could be better implemented in a service that stores the capabilities at the beginning and reloads the session in the `suite:end` hook.

**Comment by aboutlo** [comment]
@christian-bromann why I can't simply add in wdio.conf :
```
afterScenario: function (scenario) {
    browser.session('delete');
  }
```

I got 
```
CommandError: Don't end the session manually. This will be done automatically.
    at session("delete") - index.js:293:37
```
from https://github.com/webdriverio/webdriverio/blob/master/lib/protocol/session.js#L61

**Comment by christian-bromann** [comment]
Because people got confused how to run testrunner. I don't really want that someone can start and end sessions as they like. This leads to more confusion and can break reporters and services. I am open to have a openfin service!

**Comment by andyboyne** [comment]
At the moment I have worked around this by injecting location.reload() in the afterTest() hook and tearing the container down only in the afterSuite() hook (in preparation of the new session).  I'm starting to think it's not so bad as it actually makes it quicker during a suite to just reload the page than tear it down/recreate etc.

**Comment by no81no** [comment]
+1 for ability to choose if the given test in the same file should open a new browser instance or not.

**Comment by origaut** [comment]
+1 for the ability to open a new browser instance in a test within a file.

**Comment by hdattada** [comment]
+1 to reconnect the browser when the connection is lost in the middle of a test

**Comment by amyjen** [comment]
+1 for the ability to manually control my browser sessions between tests within a file

**Comment by rgustavsson** [comment]
I was aiming for having one login.feature file with multiple scenarios, e.g. 

Scenario: Successful login
Scenario: Invalid username
Scenario: Invalid password

The problem here is that Scenario 2 would have to log out from the application. I was simply expecteing Scenario 2 & 3 two be independant of Scenario 1.

+1 for the ability to open a new browser instance in a test within a file.

**Comment by christian-bromann** [comment]
@rgustavsson I would recommend to first test the failing login attempts and the successful one at the end.

**Comment by mj1618** [comment]
I would also like this functionality to destroy session, it is much better to keep tests discrete as each test should test one thing, and should be repeatable regardless of the tests that ran before it.

**Comment by christian-bromann** [comment]
> as each test should test one thing

Right, so you can test that one thing in one spec file and test another thing in a new spec file!

You don't need a new session everytime you test something. I believe it is a good compromise to say one spec file contains one selenium session. It isn't bad to create multiple files.

**Comment by mj1618** [comment]
For those of us who have hundreds of tests it is convenient to group similar tests into a file... I want to avoid having hundreds of files.

**Comment by christian-bromann** [comment]
> I want to avoid having hundreds of files.

I agree, but it is also not good to create a new session for each every small test. That would increase the test execution time and is unnecessary as in most cases a page reload is sufficient enough to start fresh again.

**Comment by mj1618** [comment]
I think very rarely is a page load sufficient to start a fresh in any stateful web application.

**Comment by christian-bromann** [comment]
> I think very rarely is a page load sufficient to start a fresh in any stateful web application.

Why? I am testing stateful applications all the time. I don't need to relogin with a new user for each feature I test on the page. For example if I test a blog application, I don't create a new user and blog post every time I want to check if the upvote or edit feature works. I do that in one session.

**Comment by mj1618** [comment]
I guess it's just a matter of preference, and would be nice to have the flexibility.

**Comment by christian-bromann** [comment]
Ok, I've implemented a `refresh` command that closes the current session and opens a new one with the same capabilities. Will release it as `v4.2.0`

**Comment by monolithed** [comment]
I don't understand why someone wants to open a new browser instance in a test within a file.
Maybe you should reconsider you workflow. I've an issue in the JIRA tracker and 1 issue === 1 test == 1 file and it works fine with 4.5k tests

**Comment by mj1618** [comment]
Thank you Christian, appreciate that. Maybe not every test file monolithed but I do have cases where it would be quite convenient. Maybe you're right maybe I need to rethink my workflow, but appreciate the flexibility to reset the session anyway. Cheers

**Comment by aboutlo** [comment]
Great news, very useful with cucumber scenario within the same file

**Comment by yuxli066** [comment]
In the hook "afterScenario: function (scenario)", how can I check that it is the last scenario in the file that I am running so that I do not have to refresh the session one more time by using "browser.reload". It's not causing a huge issue to my tests but I would prefer to be able to not waste the extra second having it refresh even after the very last scenario.

**Comment by pligor** [comment]
I believe the `reload` mentioned a bit above by @christian-bromann  is really the `reloadSession` nowadays or not ?

**Comment by christian-bromann** [comment]
> is really the `reloadSession` nowadays or not ?

Yes.

**Comment by normanlei0901** [comment]
> > is really the `reloadSession` nowadays or not ?
> 
> Yes.

@christian-bromann  question on `reload`, I see the `reload` did reload the browser instance for us. But within Browserstack, it won't create a new session within a Browserstack build but different spec file did. Thanks

**Comment by Mansouri147** [comment]
The reloadSession changes nothing in browserstack I'll have a look what I can achieve from the after hooks

**Comment by DhanalakshmiRamasamy** [comment]
> In the hook "afterScenario: function (scenario)", how can I check that it is the last scenario in the file that I am running so that I do not have to refresh the session one more time by using "browser.reload". It's not causing a huge issue to my tests but I would prefer to be able to not waste the extra second having it refresh even after the very last scenario.

@yuxli066  - Did you figure out a solution to this one ?
